content/posts/faster-rfft-python/index.md

```diff
@@ -0,0 +1,276 @@
+
+As shown in the performance plot above, using `pyfftw.FFTW` directly resulted in a significant performance improvement compared to the previous version. We achieved about 2.5x speed-up for short-size arrays (arrays with lengths `<2^7`), and around 1.5x speed-up for larger arrays, where the length of arrays is `>2^15`. For array sizes in between, we observed that the speed-up is less than 25% for most cases. 
+
+### Third Attempt: Reuse the RFFT object (V3)
+In the previous version, we intentionally created the RFFT object inside the `__call__` method. This means that for every call, a new RFFT object is created. The goal here is to see how much performance gain we can get by reusing the RFFT object across multiple calls when the input size does not change.
+
+```python
+class rfft_caller_v3:
+    def __init__(self):
+        self.real_arr = None
+        self.complex_arr = None
+        self.rfft_obj = None
+    
+    
+    def __call__(self, T):
+        if self.real_arr is None or len(T) != len(self.real_arr):
+            self.real_arr = pyfftw.empty_aligned(len(T), dtype='float64')    
+            self.complex_arr = pyfftw.empty_aligned(len(T) // 2 + 1, dtype='complex128')
+
+            self.rfft_obj = pyfftw.FFTW(
+                    self.real_arr,
+                    self.complex_arr,
+                    direction='FFTW_FORWARD',
+                    flags=('FFTW_MEASURE', 'FFTW_DESTROY_INPUT'),
+                    threads=1,
+                ) 
+           
+        self.real_arr[:] = T
+        self.rfft_obj.execute()
+
+        return self.complex_arr
+```
+
+Let's update the dictionary `rfft_callers` to include this new version:
+
+```python
+rfft_callers = {
+    'V0': rfft_caller_v0,
+    'V1': rfft_caller_v1(),
+    'V2': rfft_caller_v2(),
+    'V3': rfft_caller_v3(),
+}
+```
+
+![Performance Gain](Figure_Performance_V3.png)
+
+As shown in the performance plot above, reusing the RFFT object resulted in a huge performance improvement compared to the previous versions. We achieved about 30x speed-up for arrays with lengths `<2^8`. And from that point on, the speed-up gradually decreases to around 2x-3x for larger arrays`. This shows the overhead of creating RFFT object is considerable even when plan (wisdom) is already available. 
+
+If the goal is to compute RFFT on multiple arrays with the same size, reusing the RFFT object is straightforward. You only create it once (in which the plan is computed as well), and then you can use it multiple times without any overhead of creating the RFFT object again. However, if the input sizes vary a lot, and you need to compute RFFT on arrays with a certain, different sizes sequentially, and do that again and again, then caching the RFFT objects for different sizes can significantly improve performance. However, this requires more work to implement the caching mechanism, and it requires more memory to store the cached RFFT objects. This is outside the scope of this post.
+
+## Conclusion
```

Comment:
I would add a bonus section AFTER the conclusion where you document things that DIDN'T work to help improve performance. It would be super useful